## Table Of Contents
* [1- Loading the Dataset](#1-LoadingtheDataset)
* [2- Data-Preprocessing](#Data-Preprocessing)

<h1>1- Loading the Dataset </h1>

In [2]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv(r'dataset//cell2celltrain.csv', delimiter = ",",  header = 0, index_col = 0)

In [4]:
data.shape


(51047, 57)

In [5]:
data.info()


<class 'pandas.core.frame.DataFrame'>
Index: 51047 entries, 3000002 to 3399994
Data columns (total 57 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Churn                      51047 non-null  object 
 1   MonthlyRevenue             50891 non-null  float64
 2   MonthlyMinutes             50891 non-null  float64
 3   TotalRecurringCharge       50891 non-null  float64
 4   DirectorAssistedCalls      50891 non-null  float64
 5   OverageMinutes             50891 non-null  float64
 6   RoamingCalls               50891 non-null  float64
 7   PercChangeMinutes          50680 non-null  float64
 8   PercChangeRevenues         50680 non-null  float64
 9   DroppedCalls               51047 non-null  float64
 10  BlockedCalls               51047 non-null  float64
 11  UnansweredCalls            51047 non-null  float64
 12  CustomerCareCalls          51047 non-null  float64
 13  ThreewayCalls              51047 non-null  

In [6]:
data.head()

Churn  MonthlyRevenue  MonthlyMinutes  TotalRecurringCharge  \
CustomerID                                                               
3000002      Yes           24.00           219.0                  22.0   
3000010      Yes           16.99            10.0                  17.0   
3000014       No           38.00             8.0                  38.0   
3000022       No           82.28          1312.0                  75.0   
3000026      Yes           17.14             0.0                  17.0   

            DirectorAssistedCalls  OverageMinutes  RoamingCalls  \
CustomerID                                                        
3000002                      0.25             0.0           0.0   
3000010                      0.00             0.0           0.0   
3000014                      0.00             0.0           0.0   
3000022                      1.24             0.0           0.0   
3000026                      0.00             0.0           0.0   

            PercChangeMinutes  PercChangeRevenues  DroppedCalls  ...  \
CustomerID                                                       ...   
3000002                -157.0               -19.0           0.7  ...   
3000010                  -4.0                 0.0           0.3  ...   
3000014                  -2.0                 0.0           0.0  ...   
3000022                 157.0                 8.1          52.0  ...   
3000026                   0.0                -0.2           0.0  ...   

            ReferralsMadeBySubscriber  IncomeGroup  OwnsMotorcycle  \
CustomerID                                                           
3000002                             0            4              No   
3000010                             0            5              No   
3000014                             0            6              No   
3000022                             0            6              No   
3000026                             0            9              No   

            AdjustmentsToCreditRating  HandsetPrice  MadeCallToRetentionTeam  \
CustomerID                                                                     
3000002                             0            30                      Yes   
3000010                             0            30                       No   
3000014                             0       Unknown                       No   
3000022                             0            10                       No   
3000026                             1            10                       No   

            CreditRating  PrizmCode    Occupation  MaritalStatus  
CustomerID                                                        
3000002        1-Highest   Suburban  Professional             No  
3000010         4-Medium   Suburban  Professional            Yes  
3000014           3-Good       Town        Crafts            Yes  
3000022         4-Medium      Other         Other             No  
3000026        1-Highest      Other  Professional            Yes  

[5 rows x 57 columns]

<h1> 2- Data-Preprocessing </h1> 

<b>checking if there are rows that contain empty enteries</b>

In [7]:
rows_without_missing_data = data.dropna()

print(f"all rows: {data.shape}") 
print(f"rows without missing data: {rows_without_missing_data.shape}")
rows_w_missing_data=data.shape[0]-rows_without_missing_data.shape[0]


all rows: (51047, 57)
rows without missing data: (49752, 57)


In [8]:
rows_w_missing_data=data.shape[0]-rows_without_missing_data.shape[0]

In [9]:
rows_w_missing_data

1295

<b>checking that values of dependent variable (Churn) contains only either Yes or No  </b>

In [10]:
data['Churn'].unique()

array(['Yes', 'No'], dtype=object)

<b> Confirmed. no mistakes in data entry for this Class</b>

<h3>There are columns in the data that are of type "object" and it is unclear what their real data type is supposed to look like. </h3> </br>
<b>attempting to figure which coloumns are categorical, which are boolean, and which are strings</b>

In [15]:
for column in data.columns:
    if len(data[column].unique()) <= 10:
        print(data[column].unique())

['Yes' 'No']
['No' 'Yes']
['No' 'Yes']
['Yes' 'No']
['No' 'Yes']
['No' 'Yes']
['Known' 'Unknown']
['Yes' 'No']
['Yes' 'No']
['No' 'Yes']
['No' 'Yes']
['Yes' 'No']
['Yes' 'No']
[1 0 3 2 4]
[0 1 2 3]
['No' 'Yes']
['No' 'Yes']
[4 5 6 9 1 7 3 0 8 2]
['No' 'Yes']
['Yes' 'No']
['1-Highest' '4-Medium' '3-Good' '6-VeryLow' '2-High' '5-Low' '7-Lowest']
['Suburban' 'Town' 'Other' 'Rural']
['Professional' 'Crafts' 'Other' 'Self' 'Retired' 'Homemaker' 'Clerical'
 'Student']
['No' 'Yes' 'Unknown']


In [21]:
# select only the categorical columns
cat_df = data.select_dtypes(include='int')
# print the column names and the unique values
for col in cat_df.columns:
    print(col)
    print(cat_df[col].unique())


MonthsInService
[61 58 60 59 53 57 55 56 54 52 51 50 49 48 47 46 45 44 43 42 41 40 39 38
 37 36 35 34 33 32 29 31 30 23 24 28 12 27 26 25  8 10 22 19 21 18 20 17
 15 14 16 13 11  9  7  6]
UniqueSubs
[  2   1   3   5   6   4   7  12  11   9   8  10  13  18 196]
ActiveSubs
[ 1  2  3  4  0  5  8  6  9 53 11  7]
RetentionCalls
[1 0 3 2 4]
RetentionOffersAccepted
[0 1 2 3]
ReferralsMadeBySubscriber
[ 0  1  2  3  4  7  5  6 16 11  8  9 35]
IncomeGroup
[4 5 6 9 1 7 3 0 8 2]
AdjustmentsToCreditRating
[ 0  1  3  2  5  4  9  6 10  8  7 11 25 21 12]


In [22]:
import pandas as pd

# apply the type function to each element
types = data.apply(lambda x: x.apply(type))
# get the unique types for each column
for col in types.columns:
    print(col)
    print(types[col].unique())


Churn
[<class 'str'>]
MonthlyRevenue
[<class 'float'>]
MonthlyMinutes
[<class 'float'>]
TotalRecurringCharge
[<class 'float'>]
DirectorAssistedCalls
[<class 'float'>]
OverageMinutes
[<class 'float'>]
RoamingCalls
[<class 'float'>]
PercChangeMinutes
[<class 'float'>]
PercChangeRevenues
[<class 'float'>]
DroppedCalls
[<class 'float'>]
BlockedCalls
[<class 'float'>]
UnansweredCalls
[<class 'float'>]
CustomerCareCalls
[<class 'float'>]
ThreewayCalls
[<class 'float'>]
ReceivedCalls
[<class 'float'>]
OutboundCalls
[<class 'float'>]
InboundCalls
[<class 'float'>]
PeakCallsInOut
[<class 'float'>]
OffPeakCallsInOut
[<class 'float'>]
DroppedBlockedCalls
[<class 'float'>]
CallForwardingCalls
[<class 'float'>]
CallWaitingCalls
[<class 'float'>]
MonthsInService
[<class 'int'>]
UniqueSubs
[<class 'int'>]
ActiveSubs
[<class 'int'>]
ServiceArea
[<class 'str'> <class 'float'>]
Handsets
[<class 'float'>]
HandsetModels
[<class 'float'>]
CurrentEquipmentDays
[<class 'float'>]
AgeHH1
[<class 'float'>]
AgeH

In [12]:
data.select_dtypes(include='object')

Churn ServiceArea ChildrenInHH HandsetRefurbished  \
CustomerID                                                     
3000002      Yes   SEAPOR503           No                 No   
3000010      Yes   PITHOM412          Yes                 No   
3000014       No   MILMIL414          Yes                 No   
3000022       No   PITHOM412           No                 No   
3000026      Yes   OKCTUL918           No                 No   
...          ...         ...          ...                ...   
3399958      Yes   LAXSFN818           No                Yes   
3399974       No   LAXCDG310          Yes                 No   
3399978      Yes   LAXCDG310           No                 No   
3399990       No   NEVPOW619          Yes                 No   
3399994       No   NEVPOW619           No                 No   

           HandsetWebCapable TruckOwner RVOwner Homeownership  \
CustomerID                                                      
3000002                  Yes         No      No         Known   
3000010                   No         No      No         Known   
3000014                   No         No      No       Unknown   
3000022                  Yes         No      No         Known   
3000026                   No         No      No         Known   
...                      ...        ...     ...           ...   
3399958                  Yes         No      No         Known   
3399974                  Yes         No      No         Known   
3399978                  Yes         No      No         Known   
3399990                  Yes         No      No       Unknown   
3399994                  Yes         No      No       Unknown   

           BuysViaMailOrder RespondsToMailOffers  ... HasCreditCard  \
CustomerID                                        ...                 
3000002                 Yes                  Yes  ...           Yes   
3000010                 Yes                  Yes  ...           Yes   
3000014                  No                   No  ...           Yes   
3000022                 Yes                  Yes  ...           Yes   
3000026                 Yes                  Yes  ...           Yes   
...                     ...                  ...  ...           ...   
3399958                 Yes                  Yes  ...           Yes   
3399974                 Yes                  Yes  ...           Yes   
3399978                  No                   No  ...           Yes   
3399990                  No                   No  ...            No   
3399994                  No                   No  ...            No   

           NewCellphoneUser NotNewCellphoneUser OwnsMotorcycle HandsetPrice  \
CustomerID                                                                    
3000002                  No                  No             No           30   
3000010                 Yes                  No             No           30   
3000014                 Yes                  No             No      Unknown   
3000022                 Yes                  No             No           10   
3000026                  No                 Yes             No           10   
...                     ...                 ...            ...          ...   
3399958                  No                  No             No           60   
3399974                  No                  No             No           60   
3399978                  No                  No             No           80   
3399990                  No                  No             No           30   
3399994                  No                  No             No           60   

           MadeCallToRetentionTeam CreditRating PrizmCode    Occupation  \
CustomerID                                                                
3000002                        Yes    1-Highest  Suburban  Professional   
3000010                         No     4-Medium  Suburban  Professional   
3000014                         No       3-Good      Town        Crafts   
30000

<b>returning all columns of type object</b>

In [30]:
#selecting columns with type 'object'

data = data.select_dtypes(include='object')

obj_data_subset.sample(5)

Churn ServiceArea ChildrenInHH HandsetRefurbished  \
CustomerID                                                     
3367354       No   ATHJHC423           No                 No   
3210314       No   NYCMAN917           No                 No   
3155350       No   AIRGRE864          Yes                 No   
3281994       No   LAXVNY818           No                 No   
3006618       No   SFRSMO650           No                Yes   

           HandsetWebCapable TruckOwner RVOwner Homeownership  \
CustomerID                                                      
3367354                  Yes         No      No       Unknown   
3210314                  Yes         No      No         Known   
3155350                  Yes         No      No         Known   
3281994                  Yes         No      No       Unknown   
3006618                  Yes         No      No         Known   

           BuysViaMailOrder RespondsToMailOffers  ... HasCreditCard  \
CustomerID                                        ...                 
3367354                  No                   No  ...            No   
3210314                  No                   No  ...           Yes   
3155350                  No                   No  ...           Yes   
3281994                  No                   No  ...            No   
3006618                 Yes                  Yes  ...           Yes   

           NewCellphoneUser NotNewCellphoneUser OwnsMotorcycle HandsetPrice  \
CustomerID                                                                    
3367354                 Yes                  No             No      Unknown   
3210314                  No                  No             No          200   
3155350                  No                  No             No          150   
3281994                  No                  No             No           30   
3006618                 Yes                  No             No          200   

           MadeCallToRetentionTeam CreditRating PrizmCode    Occupation  \
CustomerID                                                                
3367354                         No       2-High     Rural         Other   
3210314                         No       3-Good     Other         Other   
3155350                         No     7-Lowest      Town         Other   
3281994                         No        5-Low     Other         Other   
3006618                         No    1-Highest  Suburban  Professional   

           MaritalStatus  
CustomerID                
3367354          Unknown  
3210314               No  
3155350               No  
3281994          Unknown  
3006618               No  

[5 rows x 23 columns]

In [34]:
len(obj_data_subset["ServiceArea"].unique())

748

<b>* note: we have 748 Service Areas<b> </br>
<b> Service Area dtype = String</b>

In [39]:
for column in obj_data_subset.columns:
        print(f"{column}: {len(obj_data_subset[column].unique())}")

Churn: 2
ServiceArea: 748
ChildrenInHH: 2
HandsetRefurbished: 2
HandsetWebCapable: 2
TruckOwner: 2
RVOwner: 2
Homeownership: 2
BuysViaMailOrder: 2
RespondsToMailOffers: 2
OptOutMailings: 2
NonUSTravel: 2
OwnsComputer: 2
HasCreditCard: 2
NewCellphoneUser: 2
NotNewCellphoneUser: 2
OwnsMotorcycle: 2
HandsetPrice: 16
MadeCallToRetentionTeam: 2
CreditRating: 7
PrizmCode: 4
Occupation: 8
MaritalStatus: 3


<b>features that have 2 values only are most likely boolean</b>

<b>Homeownership is boolean, but describing "Homeownership" as true for Known, and False for Unknown, does not correctly depict the meaning</b> </br>
<b>therefore, change in label iteself, as "Homeownership_Known" such that if True, it means that we know of the Homeownership,  delivering the correct meaning</b>

In [57]:
data.rename(columns={"Homeownership": "Homeownership_Known"}, inplace=True)


In [59]:
data['Homeownership_Known'].head()

CustomerID
3000002      Known
3000010      Known
3000014    Unknown
3000022      Known
3000026      Known
Name: Homeownership_Known, dtype: object

In [60]:
for column in data.columns:
    if len(data[column].unique() )== 2: 
        print(f"feature: {column} is likely boolean")
        print(f"sample: {data[column].sample(4)}")
        data[column] = data[column].str.lower().map({'yes': True, 'known': True, 'no':False, 'unknown': False})
        print(f"changed datatype: {data[column].dtype}")
        

feature: Churn is likely boolean
sample: CustomerID
3324182    No
3201774    No
3273770    No
3074834    No
Name: Churn, dtype: object
changed datatype: bool
feature: ChildrenInHH is likely boolean
sample: CustomerID
3005482    No
3138318    No
3080230    No
3035702    No
Name: ChildrenInHH, dtype: object
changed datatype: bool
feature: HandsetRefurbished is likely boolean
sample: CustomerID
3092362    No
3022626    No
3189914    No
3319850    No
Name: HandsetRefurbished, dtype: object
changed datatype: bool
feature: HandsetWebCapable is likely boolean
sample: CustomerID
3080102    Yes
3110926    Yes
3213654    Yes
3265210    Yes
Name: HandsetWebCapable, dtype: object
changed datatype: bool
feature: TruckOwner is likely boolean
sample: CustomerID
3133610     No
3337870    Yes
3209482     No
3110118     No
Name: TruckOwner, dtype: object
changed datatype: bool
feature: RVOwner is likely boolean
sample: CustomerID
3331174     No
3164042     No
3193210     No
3209110    Yes
Name: RVOwner,

In [63]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51047 entries, 3000002 to 3399994
Data columns (total 57 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Churn                      51047 non-null  bool   
 1   MonthlyRevenue             50891 non-null  float64
 2   MonthlyMinutes             50891 non-null  float64
 3   TotalRecurringCharge       50891 non-null  float64
 4   DirectorAssistedCalls      50891 non-null  float64
 5   OverageMinutes             50891 non-null  float64
 6   RoamingCalls               50891 non-null  float64
 7   PercChangeMinutes          50680 non-null  float64
 8   PercChangeRevenues         50680 non-null  float64
 9   DroppedCalls               51047 non-null  float64
 10  BlockedCalls               51047 non-null  float64
 11  UnansweredCalls            51047 non-null  float64
 12  CustomerCareCalls          51047 non-null  float64
 13  ThreewayCalls              51047 non-null  

<b>that way, we changed all boolean objects to their correct data type</b>

<b> regarding MaritalStatus, changing its datatype to category, where unknown category is Nan</b>

In [64]:
data['MaritalStatus'].unique()

array(['No', 'Yes', 'Unknown'], dtype=object)

In [68]:
data['MaritalStatus'].str.lower().replace('unknown', np.nan, inplace=True)

data['MaritalStatus'] = data['MaritalStatus'].astype('category')


In [69]:
data['MaritalStatus'].info()

<class 'pandas.core.series.Series'>
Index: 51047 entries, 3000002 to 3399994
Series name: MaritalStatus
Non-Null Count  Dtype   
--------------  -----   
51047 non-null  category
dtypes: category(1)
memory usage: 448.8 KB


In [71]:
data['MaritalStatus'].sample(10)

CustomerID
3317674    Unknown
3115366        Yes
3328014        Yes
3278106    Unknown
3373118         No
3256878         No
3330710        Yes
3126414        Yes
3251238        Yes
3304406    Unknown
Name: MaritalStatus, dtype: category
Categories (3, object): ['No', 'Unknown', 'Yes']

In [100]:
for column in data.columns:
    if data.dtypes[column] == 'object':
        print(f"{column}: {len(data[column].unique())}")
        print(f"sample: {data[column].unique()}")
        print("----------------------------------------------------------------")

ServiceArea: 748
sample: ['SEAPOR503' 'PITHOM412' 'MILMIL414' 'OKCTUL918' 'OKCOKC405' 'SANMCA210'
 'SLCSLC801' 'LOULOU502' 'KCYKCK913' 'KCYNEW316' 'KCYKCM816' 'DENDEN303'
 'PHICTR610' 'OKCLRK501' 'OMADES515' 'SANAUS512' 'KCYWIC316' 'INDIND317'
 'SLCPRO801' 'OMALNC402' 'NSHNSH615' 'OMAOMA402' 'PHXTUC520' 'DALDAL214'
 'DALDTN940' 'DALFTW817' 'SANSAN210' 'NYCMAN917' 'NOLKEN504' 'MINMIN612'
 'NNYROC716' 'PHIARD610' 'NYCQUE917' 'DENBOU303' 'BOSBOS617' 'PHXPHX602'
 'NNYBUF716' 'DETTOL419' 'NNYSYR315' 'STLSTL314' 'PHIPHI215' 'SEABLV425'
 'MIAMIA305' 'PHICHC215' 'SEASEA206' 'BIRBIR205' 'MIADFD954' 'LOUNAL812'
 'MIADEL561' 'MIAFTL954' 'SEASPO509' 'SEATAC253' 'SFRSAC916' 'MIANDA305'
 'MINSTP612' 'MIAHWD954' 'NYCBRO917' 'PHIJEN215' 'OMACDR319' 'MIAWPB561'
 'SEAOLY360' 'PITBUT412' 'SEAEVE425' 'PHIMER609' 'SEACDA208' 'HARHAR860'
 'SFRSFR415' 'SFROAK510' 'SFRSCL408' 'NNYALB518' 'STLCOL618' 'NYCNEW201'
 'SFRPAL650' 'DETDET313' 'DETROS810' 'NYCNEW908' 'NYCNAS516' 'NSHCOL615'
 'PITGRE412' 'SEABEA503' '

In [75]:
#changing the features that have finite set of values to categorical

data['PrizmCode'] = data['PrizmCode'].astype('category')
data['Occupation'] = data['Occupation'].astype('category')


<b>CreditRating feature seems to be ordinal data</b>


since we still dont know exactly if this feature will be used later on and how exactly, seems viable to separate the creditrating feature into 2, one is numerical, and the other is string (categorical)

In [78]:
#Extract numerical part
data['CreditRating_Numeric'] = data['CreditRating'].str[:1] 

#Extract string part
data['CreditRating_Label'] = data['CreditRating'].str[2:] 


In [81]:
data['CreditRating_Numeric'] =data['CreditRating_Numeric'].astype('int64')
data['CreditRating_Label'] = data['CreditRating_Label'].astype('category')

drop the CreditRating Column

In [84]:
data.drop(columns=['CreditRating'], inplace=True)


<b>convert HandsetPrice to float, and Unknown values will be 'NaN'</b>


In [82]:
data['HandsetPrice'] = pd.to_numeric(data['HandsetPrice'], errors='coerce')

<b>ServiceArea has 748 unique values, its not very practical to one-hot encode it. </b>

In [88]:
#extract the string part
data['City'] = data['ServiceArea'].str[:-3]
#extract the last three digits
data['AreaCode'] = data['ServiceArea'].str[-3:]


In [92]:
len(data['City'].unique())

737

In [93]:
len(data['AreaCode'].unique())

197

<b>attempt to perform frequency encoding </b>

In [101]:
city_counts = data['City'].value_counts()
data['SerivceArea_City_Frequency'] = data['City'].map(city_counts)

area_code_counts = data['AreaCode'].value_counts()
data['ServiceArea_AreaCode_Frequency'] = data['AreaCode'].map(area_code_counts)


In [105]:

data.drop(columns=['City_Frequency'], inplace=True)
data.drop(columns=['AreaCode_Frequency'], inplace=True)


In [106]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51047 entries, 3000002 to 3399994
Data columns (total 59 columns):
 #   Column                          Non-Null Count  Dtype   
---  ------                          --------------  -----   
 0   Churn                           51047 non-null  bool    
 1   MonthlyRevenue                  50891 non-null  float64 
 2   MonthlyMinutes                  50891 non-null  float64 
 3   TotalRecurringCharge            50891 non-null  float64 
 4   DirectorAssistedCalls           50891 non-null  float64 
 5   OverageMinutes                  50891 non-null  float64 
 6   RoamingCalls                    50891 non-null  float64 
 7   PercChangeMinutes               50680 non-null  float64 
 8   PercChangeRevenues              50680 non-null  float64 
 9   DroppedCalls                    51047 non-null  float64 
 10  BlockedCalls                    51047 non-null  float64 
 11  UnansweredCalls                 51047 non-null  float64 
 12  CustomerCareCal

In [107]:
for column in data.columns:
    if data.dtypes[column] == 'object':
        print(f"{column}: {len(data[column].unique())}")
        print(f"sample: {data[column].unique()}")
        print("----------------------------------------------------------------")